### PR TITLE
#159 - light-dark-mode

### PR DESCRIPTION
**Pull request by jevonmao**
@gao-sun I added a toggle to switch between dark and light mode. You want to take a look before merging?

### CODE REVIEW

**Comment by gao-sun** [comment]
taking a look tonight

File: eul/Base.lproj/Main.storyboard [hunk 1]
```diff
@@ -1,7 +1,8 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<document type="com.apple.InterfaceBuilder3.Cocoa.Storyboard.XIB" version="3.0" toolsVersion="14814" targetRuntime="MacOSX.Cocoa" propertyAccessControl="none" useAutolayout="YES">
```
Comment (gao-sun) [inline eul/Base.lproj/Main.storyboard:2]: what's this file change for?

File: eul/Views/Preference/PreferenceDisplayView.swift [hunk 2]
```diff
@@ -49,6 +49,28 @@ extension Preference {
...
+
+                } else {
+                    Picker("appearance.mode".localized(), selection: $preference.appearanceMode) {
+                        ForEach(appearanceMode) {
+                            Text($0.description)
+                                .tag($0)
+                        }
+                    }
+                    .frame(width: 200)
+                }
```
Comment (gao-sun) [inline eul/Views/Preference/PreferenceDisplayView.swift:73]: what's the difference if we don't put `onChange` in the end?

Comment (jevonmao) [reply eul/Views/Preference/PreferenceDisplayView.swift:73]: @gao-sun Do you know a way to respond to changes of a `Published` variable in the view model? I could not figure it out so I had to use `onChange`. The `onChange` detects if `appearanceMode` has changed, and calls the function to alter dark and light mode.

Comment (gao-sun) [reply eul/Views/Preference/PreferenceDisplayView.swift:73]: i see. you can use `didSet` for listening changes of `@Published` in view model

**Comment by gao-sun** [comment]
thanks for taking my advices, will take a look tonight

File: eul/Schema/Preference.swift [hunk 3]
```diff
@@ -23,6 +23,25 @@ struct Preference {
...
+            case .auto:
+                return "appearance.auto".localized()
+
+            case .dark:
+                return "appearance.dark".localized()
+
+            case .light:
+                return "appearance.light".localized()
+            }
+        }
```
Comment (gao-sun) [inline eul/Schema/Preference.swift:42]: could be `"appeareance.\(rawValue)".localized()`

File: eul/StatusBar/StatusBarItem.swift [hunk 4]
```diff
@@ -102,6 +112,11 @@ class StatusBarItem: NSObject, NSMenuDelegate {
         super.init()
 
         statusBarMenu.delegate = self
+        if preferenceStore.appearanceMode == .light {
+            statusBarMenu.appearance = NSAppearance(named: .aqua)
+        } else {
+            statusBarMenu.appearance = NSAppearance(named: .darkAqua)
+        }
```
Comment (gao-sun) [inline eul/StatusBar/StatusBarItem.swift:119]: maybe re-use `changeNSWindowColorScheme`?

File: eul/Store/PreferenceStore.swift [hunk 5]
```diff
@@ -58,6 +58,11 @@ class PreferenceStore: ObservableObject {
     @Published var checkStatusItemVisibility = true
     @Published var isUpdateAvailable: Bool? = false
     @Published var checkUpdateFailed = true
+    @Published var appearanceMode = Preference.appearance.auto {
+        didSet {
+            changeColorScheme()
+        }
+    }
```
Comment (gao-sun) [inline eul/Store/PreferenceStore.swift:65]: maybe we can leverage Combine to listen to the changes of `preferenceStore.appearanceMode` in `StatusBarManager`? like what i did with `activeCancellable`

File: eul/Views/Menu/StatusMenuView.swift [hunk 6]
```diff
@@ -15,6 +15,18 @@ struct StatusMenuView: SizeChangeView {
...
+            base
+                .environment(\.colorScheme, .light)
+        } else if preferenceStore.appearanceMode == .dark {
+            base
+                .environment(\.colorScheme, .dark)
+        } else {
+            base
+        }
+    }
+
```
Comment (gao-sun) [inline eul/Views/Menu/StatusMenuView.swift:28]: something like 

```swift
if let appearanceMode = preferenceStore.appearanceMode {
    base.environment(...)
} else {
    base
}
```

Comment (gao-sun) [reply eul/Views/Menu/StatusMenuView.swift:28]: I think it's a good scenario for using `ViewModifier`

File: eul/StatusBar/StatusBarItem.swift (hunk 4 above)
Comment (jevonmao) [reply eul/StatusBar/StatusBarItem.swift:119]: You mean like a ternary operator for the parameter like 
`NSAppearance(named: preferenceStore.appearanceMode == .light ? .aqua : .darkAqua)`. Is this okay?

Comment (gao-sun) [reply eul/StatusBar/StatusBarItem.swift:119]: I mean the function itself - like you can call `changeNSWindowColorScheme()` here

File: eul/Views/Menu/StatusMenuView.swift (hunk 6 above)
Comment (jevonmao) [reply eul/Views/Menu/StatusMenuView.swift:28]: Sorry but I don't quite understand? What do you mean by `ViewModifier`? I used `.environment(\.colorScheme` which is a built in modifier to change SwiftUI to dark mode. Unless you recommend something else.

File: eul/StatusBar/StatusBarItem.swift (hunk 4 above)
Comment (jevonmao) [reply eul/StatusBar/StatusBarItem.swift:119]: Well the `changeNSWindowColorScheme` takes in a `NSAppearance`, so I added another function to read from preferences, and call `changeNSWindowColorScheme` with correct parameters

File: eul/Schema/Preference.swift (hunk 3 above)
Comment (jevonmao) [reply eul/Schema/Preference.swift:42]: Fixed

File: eul/Store/PreferenceStore.swift (hunk 5 above)
Comment (jevonmao) [reply eul/Store/PreferenceStore.swift:65]: Fixed

**Review by gao-sun — APPROVED** [review]
looks neat, thanks

**Pull request merged**